src/libolm_migration.rs

```diff
@@ -0,0 +1,446 @@
+impl BaseMigrationData {
+    /// Create a new `BaseMigrationData` with default values.
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+#[wasm_bindgen]
+impl Migration {
+    /// Import the base dataset from a libolm-based setup to a vodozemac-based
+    /// setup stored in IndexedDB.
+    ///
+    /// Populates the user credentials, Olm account, backup data, etc. This is the
+    /// first step in the migration process. Once this base data is imported,
+    /// further data can be imported with {@link #migrateOlmSessions}, {@link
+    /// #migrateMegolmSessions}, and TODO room settings.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - The data to be migrated
+    /// * `pickle_key` - The libolm pickle key that was used to pickle the olm
+    ///   account objects.
+    /// * `store_handle` - A connection to the CryptoStore which will be used to
+    ///   store the vodozemac data.
+    #[wasm_bindgen(js_name="migrateBaseData")]
+    pub async fn migrate_base_data(
+        data: &BaseMigrationData,
+        pickle_key: Uint8Array,
+        store_handle: &StoreHandle,
+    ) -> Result<JsValue, JsError> {
+        migrate_base_data_to_store(data, &(pickle_key.to_vec()), store_handle.store.as_ref())
+            .await
+            .map_err(|e| JsError::from(&*e))?;
+        Ok(JsValue::UNDEFINED)
+    }
+}
+
+async fn migrate_base_data_to_store(
+    data: &BaseMigrationData,
+    pickle_key: &[u8],
+    store: &DynCryptoStore,
+) -> anyhow::Result<()> {
+    let user_id = data.user_id.clone().context("User ID not specified")?.inner;
+    let account = vodozemac::olm::Account::from_libolm_pickle(&data.pickled_account, pickle_key)?;
+    let account =
+        matrix_sdk_crypto::olm::Account::from_pickle(matrix_sdk_crypto::olm::PickledAccount {
+            user_id: user_id.clone(),
+            device_id: data.device_id.clone().context("Device ID not specified")?.inner,
+            pickle: account.pickle(),
+            // Legacy crypto in the js-sdk does not keep a record of whether it has published the
+            // device keys to the server (it does it every time the stack is started). For safety,
+            // let's assume it hasn't happened yet.
+            shared: false,
+            // Assume we have 50 keys on the server, until we get a sync that says fewer.
+            uploaded_signed_key_count: 50,
+            creation_local_time: MilliSecondsSinceUnixEpoch::now(),
+        })?;
+
+    let backup_decryption_key = data
+        .backup_recovery_key.as_ref()
+        .map(|k| BackupDecryptionKey::from_base64(k.as_str()))
+        .transpose()?;
+
+    let cross_signing = PrivateCrossSigningIdentity::empty(&user_id);
+    cross_signing
+        .import_secrets_unchecked(
+            data.cross_signing_master_key.as_deref(),
+            data.cross_signing_self_signing_key.as_deref(),
+            data.cross_signing_user_signing_key.as_deref(),
+        )
```

Comment:
Maybe we should `.zeroize()` the `data.cross_signing_*` at this point?

Reply:
I wondered about that. I concluded it's not worthwhile, because even if we zeroize it on the Rust side, there may well still be copies kicking around in memory from the javascript side which we have no hope of zeroizing. (As https://github.com/veorq/cryptocoding#clean-memory-of-secret-data says: "there's virtually no way to reliably clean secret data in garbage-collected languages (such as ... Javascript)").

And we don't generally zeroize any other sensitive data that is passed through the wasm bindings.